ollama/_utils.py

```diff
@@ -0,0 +1,87 @@
+from __future__ import annotations
+from collections import defaultdict
+import inspect
+from typing import Callable, Union
+import re
+
+import pydantic
+from ollama._types import Tool
+
+
+def _parse_docstring(doc_string: Union[str, None]) -> dict[str, str]:
+  parsed_docstring = defaultdict(str)
+  if not doc_string:
+    return parsed_docstring
+
+  key = hash(doc_string)
+  for line in doc_string.splitlines():
+    lowered_line = line.lower().strip()
+    if lowered_line.startswith('args:'):
+      key = 'args'
+    elif lowered_line.startswith('returns:') or lowered_line.startswith('yields:') or lowered_line.startswith('raises:'):
+      key = '_'
+
+    else:
+      # maybe change to a list and join later
+      parsed_docstring[key] += f'{line.strip()}\n'
+
+  last_key = None
+  for line in parsed_docstring['args'].splitlines():
+    line = line.strip()
+    if ':' in line:
+      # Split the line on either:
+      # 1. A parenthetical expression like (integer) - captured in group 1
+      # 2. A colon :
+      # Followed by optional whitespace. Only split on first occurrence.
+      parts = re.split(r'(?:\(([^)]*)\)|:)\s*', line, maxsplit=1)
```

Comment:
I tend to avoid regexp when possible since it's hard to grok. In this scenario, a simpler solution would be to split on the mandatory `:` than parse the pre-colon and post-colon sections independently. Here's an example that passes your tests

```python
  for line in parsed_docstring['args'].splitlines():
    pre, _, post = line.partition(':')
    if not pre.strip():
      continue
    if not post.strip() and last_key:
      parsed_docstring[last_key] += ' ' + pre
      continue

    arg_name, _, _ = pre.replace('(', ' ').partition(' ')
    last_key = arg_name.strip()

    parsed_docstring[last_key] = post.strip()
```

Reply:
TODO @ParthSareen to spin out issue